### PR TITLE
Add support for archive dir 

### PR DESCRIPTION
**Pull request by pditommaso**
This feature allows exporting all published files to a bucket path specified via NXF_ARCHIVE_DIR env variable. 

The variable value specifies source path and the target bucket path separate with a comma character.

### CODE REVIEW

**Comment by jordeu** [comment]
I didn't had time to test it, but from reading the code I see that only `.command.*` files are archived and not the ones from the workdir root. 

I mean all these files that are also required by Tower:
- `nf-<id>.log`
- `nf-<id>.txt`
- `timeline-<id>.html`
- `nf-<id>-reports.tsv`
- `nf-<id>-tower.yml`

File: modules/nextflow/src/main/groovy/nextflow/util/FileArchiver.groovy [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    boolean asBoolean() {
+        return baseDir!=null && targetDir!=null
+    }
+
+    FileArchiver() {
+        this(System.getenv())
+    }
+
+    FileArchiver(Map<String,String> env) {
+        final paths = env.get('NXF_ARCHIVE_DIR')?.tokenize(',')
```
Comment (jordeu) [inline modules/nextflow/src/main/groovy/nextflow/util/FileArchiver.groovy:50]: A `,` is a valid path character. I known that we'll have the same problem with any character but may be we can use two characters to do the splitting and reduce the probabilities that is part of the path. Also a `,` seems more appropriate to separate items of a list and this is not the case. For me `==` fits slightly better in this case.

**Comment by pditommaso** [comment]
Yes, I've added [this](https://github.com/nextflow-io/nextflow/pull/3018/commits/7b8639291eb33fc5b3c099c981bb5974a339403a) to copy them on completion. However not sure if there could be a race condition, I'm the file copy is started before the writing is determined.

File: modules/nextflow/src/main/groovy/nextflow/util/FileArchiver.groovy (hunk 1 above)
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/util/FileArchiver.groovy:50]: yeah, I thought that, was a bit lazy to escape. 

> I known that we'll have the same problem with any character

Maybe no! I was not aware that has separator specific codes  https://stackoverflow.com/a/18782271/395921

**Comment by pditommaso** [comment]
Ok, rewrote this to avoid any change other than the tower plugin.

**Comment by jordeu** [comment]
First attempt to use it I got several errors. I'll try to check them later.

```
WARN: Unable to archive file: /fsx/results/step_1/MultiQC Report.html -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/98/ba8723a1940841c460cc827a4d5989/.command.log -- cause: java.nio.file.NoSuchFileException: /fsx/98/ba8723a1940841c460cc827a4d5989/.command.log
WARN: Unable to archive file: /fsx/98/ba8723a1940841c460cc827a4d5989/.command.sh -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/report.pdf -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/baits.bed -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/98/ba8723a1940841c460cc827a4d5989/.exitcode -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/98/ba8723a1940841c460cc827a4d5989/.command.err -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/transcriptome.paf -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/98/ba8723a1940841c460cc827a4d5989/.command.out -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/98/ba8723a1940841c460cc827a4d5989/.command.run -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/test.bigwig -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/nfcore_chipseq110_samplesheet_test_full_6cols.csv -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/all_sites.fas -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/SPAdesHybrid-CAPES_S7-binDepths.heatmap.png -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/transposed_report.tex -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/genome.fasta -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/genome.gff3 -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/genome.fasta.fai -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/test2.targets.tsv.gz -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/genome.dict -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/test.vcf.gz.tbi -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/execution_trace_2021-07-29_07-12-59.txt -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/test.paired_end.bam -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/test.bedgraph -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/genome.gtf -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/bin_depths_summary.tsv -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/test.vcf -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/pipeline_dag_2021-07-29_07-12-59.svg -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/taxonomy.krona.html -- cause: java.io.IOException: Failed to put data into Amazon S3 object
WARN: Unable to archive file: /fsx/results/CAPES_S7.log -- cause: java.io.IOException: Failed to put data into Amazon S3 object
```

**Comment by pditommaso** [comment]
mehhh, what's in the log?

**Comment by jordeu** [comment]
I cannot access it. I'll repeat the test later but leaving one EC2 instance always up to be able to debug it.

**Comment by jordeu** [comment]
Ok, some errors were due to a permission problem and most of the files are correctly archived. 

But there is this error several times:
```
Caused by: java.nio.file.NoSuchFileException: /fsx/cd/f6d1e153613c7f653511cb8130ebc4/.command.log
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:148)
	at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1851)
	at nextflow.file.CopyMoveHelper.copyToForeignTarget(CopyMoveHelper.java:189)
	at nextflow.file.FileHelper.copyPath(FileHelper.groovy:939)
	at io.seqera.tower.plugin.TowerArchiver$1$_run_lambda1.doCall(TowerArchiver.groovy:167)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:236)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:75)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:176)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:437)
```
That happens because the file `.command.log` do not always exist.

Also the root Tower files are not archive and there is no error reported. The Tower env variables contain only the file name, may be the absolute path it is not correctly resolved?

Nextflow log file [nf-pub-archive.log](https://github.com/nextflow-io/nextflow/files/9096008/nf-pub-archive.log)

**Comment by pditommaso** [comment]
> That happens because the file .command.log do not always exist.

Good point. This is an easy patch 

> The Tower env variables contain only the file name, may be the absolute path it is not correctly resolved?

This is strange, is just the name but it should be resolved against running directly that's where the file is expected to be located. 

https://github.com/nextflow-io/nextflow/pull/3018/files#diff-6f4dd91275ae342af957397ce64cf7ec90f1b3bc60a815438c57a9ab60f0d846R147

**Comment by pditommaso** [comment]
Merged via 👉 c234681a

**Pull request closed without merging**